implemented CycleDetectionII code in LinkedList

Akshay73c opened the pull request:
Hey maintainers, 
I've implemented the code for getting the starting node of cycle in a linked list which i think will be useful for understanding of cycle-type-questions.
I've also added the tests to check the correctness.
It'll be my first open source contribution and also in _Hacktoberfest_. Hoping a success :)

### Changes:

- [ ] Added a code to get the starting node of cycle in LinkedList

### Checklist:
I've read the CONTRIBUTING.md file and also checked the testcases using npm as asked to. 
This pr is all my own work.

appgurueu reviewed (commented):
* The code is still messy (mostly formatting-wise, the `lengthCycle` function should be defined before use for better readability, IMO). Please add comments, and give better variable names (i.e. I assume `f` and `s` are `fast` and `slow`; why do you need them when you already have `fast` and `slow`?).
* As for the tests, remove the 3 redundant comments there which just repeat test case names.

Akshay73c commented:
Dear maintainers,
Thanks for the feedback !!  I appreciate you taking the time to look over the code and provide suggestions.
I've made the following changes as suggested: 
- Added comments explaining the logic
- you're absolutely right that using f and s is redundant when I already have fast and slow pointers
- removed the redundant comments in tests
- In my opinion, for this problem the key focus is detecting the cycle start node, so I wanted to keep that logic upfront in detectCycleNode(). Defining the helper lengthCycle() function below separates it out from the main detection logic. That said, I'm open to rearranging if you feel strongly that defining lengthCycle() first improves the overall structure.

Let me know! And thanks again for the review :)

I wanted to reach out to let you know that while attempting my first contribution to this open source project, I ran into some issues with git push. (That's why so many commits) After doing some research, I was able to resolve the issues and successfully contribute my changes.
I realize this may not have been the most elegant first contribution, and I wanted to apologize in advance if I caused any inconvenience. As a first time contributor, I was still learning the contribution workflow.

Akshay73c commented:
Hey maintainers, I'd be really glad if you could review this ;)

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+/**
+ * A LinkedList based solution for Detecting a start node of the Cycle in a list.
```
I'd say "finding" a start node rather than "detecting" one. Also consider something like `@return a node on a cycle in the linked list`.

Akshay73c replied:
Yeah, I've updated the comment to reflect that we are finding the starting node, rather than detecting it. Also added a JSDoc `@returns` comment specifying that the function returns a node on a cycle in the linked list for clarity.

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+ */
+
+function detectCycleNode(head) {
+  /*
```
This comment is redundant except for the Leetcode link (which could be added above). Please use the JSDoc `@see` annotation both for the Wiki & Leetcode link.

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  if (length === 0) {
+    // If length = 0, return null
```
Redundant comment. A more useful comment might be "if there is no cycle, return `null`".

Akshay73c replied:
Yup, your suggested comment would clarify better about the null return value.

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+  return slow
+}
+
+function lengthCycle(head) {
```
This function seems overcomplicated to me. This function is only called if there is a cycle, yet there is effectively dead/unnecessary code to handle the case that there is no cycle. Also, why do you need another "tortoise-and-hare" here? If `head` already is a node on the cycle, can't you just determine the length of the cycle by following pointers until you're back to `head`? (Effectively your inner loop.)

If I'm missing something and this additional work is necessary, there should be comments justifying / explaining its purpose. It is not self-evident.

Akshay73c replied:
Thank you for pointing out the potential overcomplication. Though it appears redundant, the inner loop and both pointers are crucial for calculating the cycle length accurately. Even if the head is already on the cycle, we need both the pointers to actually go inside the cycle to determine its length. The condition you mentioned isn't meant to be explicitly checked. It automatically falls into place as both pointers move within the cycle. Also to make the 'length' function more universal or like isolated, I've kept it independent of other functions and reusable.

appgurueu replied:
> Though it appears redundant, the inner loop and both pointers are crucial for calculating the cycle length accurately.

I never challenged the necessity of the inner loop, quite the opposite.

> Even if the head is already on the cycle, we need both the pointers to actually go inside the cycle to determine its length.

How so? As said, just the inner loop should work.

> Also to make the 'length' function more universal or like isolated, I've kept it independent of other functions and reusable.

This is pointless given that it's not exported.

Even if it was exported (and tested), I don't think duplicating code and responsibilities is a good idea. Finding a node on a cycle and finding the length of the cycle are two different things. There is no need to duplicate the "finding a node on a cycle" (tortoise and hare) logic.

To make sure that my logic is correct, I simplified the `lengthCycle` function to

```javascript
// head is a node on a cycle
function lengthCycle(head) {
  // How long until we visit head again?
  let cur = head;
  let len = 0;
  do {
  	cur = cur.next;
  	len++;
  } while (cur != head);
  return len;
}
```

and as expected, the tests are still passing.

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+    return null
+  }
+
+  //Point the nodes at head
```
Redundant comment. (Also, could be shortened to `fast = slow = head` in principle.)

Akshay73c replied:
Yes. `fast = slow = head` would be better. I have made this change as well.

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+  //Point the nodes at head
+  fast = head
+  slow = head
+  // Move slow pointer ahead by 'length' of cycle times
```
The purpose of this is unclear to me. If `length` is in nodes, then `fast = slow = head` after this and this is effectively dead code. If `length` is in "edges" / "links" (one fewer than nodes), then you will get the predecessor of `head` in the cycle, but what for?

Akshay73c replied:
Here, the 'length' refers to the number of nodes to move forward the 'slow' pointer in the linked list. (while keeping the 'fast' pointer at head only ). Then we'll move both pointers one by one to get the starting node of cycle. I have attached an image to help illustrate this point and my thought process.

Akshay73c replied:
Thought Process: 
![CycleNode1](https://github.com/TheAlgorithms/JavaScript/assets/118964099/78544105-84ac-43c0-84a2-c45bd195a664)

Akshay73c replied:
![CycleNode2](https://github.com/TheAlgorithms/JavaScript/assets/118964099/a04ec2ea-b7d2-4ca9-8a66-6daabcb14b2a)

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+    length--
+  }
+
+  // Now move both pointers until they meet - this will be the start of cycle
```
Why will it be the "start" of the cycle?

Akshay73c replied:
To help illustrate my thought process, please see the attached image above. Hopefully, it'll make things more clear. Let me know if anything is fuzzy.

appgurueu replied:
Thanks for the effort, I understand it now. Maybe the comments could do a better job conveying this?

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,77 @@
+
+  // Now move both pointers until they meet - this will be the start of cycle
+  while (fast !== slow) {
+    fast = fast.next
```
This would have to be `fast.next.next`. If this is not failing the tests in its current form, it means `fast == slow`, and all this code is practically dead.

Akshay73c replied:
I appreciate you pointing out the potential for misinterpretation. Here, the `fast` pointer is kept stationery at the head only and `slow` is moved ahead first in the list. And then we move both one by one to make them meet at the starting node. Please see the attached images above for my justification.

appgurueu replied:
Yeah, I see why now. Again, see my point about the confusing naming of these variables.

appgurueu reviewed (changes requested):
Of course! This seems to work, but there is definitely dead / unnecessary code, and why the code is doing what it is doing is not exactly clear. Also some minor stylistic remarks.

Akshay73c commented:
Dear maintainers,
Thanks for reviewing again! I appreciate you taking this much time to provide suggestions for a beginner like me. 
I've made few minor changes to the code and also explained my thought process and addressed to your valuable reviews. Please let me know if anything is required. 
And am really sorry for the delays as i was occupied with campus interviews and exams. (am looking for my first internship)

Akshay73c commented:
Please let me know if there's anything more that needs to be addressed. Fingers crossed for the merge ;)

appgurueu commented on Data-Structures/Linked-List/CycleDetectionII.js:
```diff
@@ -0,0 +1,71 @@
+    return null
+  }
+
+  fast = slow = head
```
This is what confused me. Please don't reuse these variables for something entirely different...

Given that the pointers move at the same speed, but one is "ahead" of the other (until they meet), maybe call these "ahead" and "behind"? The names "slow" and "fast" are misleading.

Akshay73c commented:
> Sorry for the long delay. I simplified the code a bit and renamed some variables. I hope you're fine with this, otherwise feel free to complain :)
> 
> LGTM!

LGTM too! Thanks for simplifying things, and no complaints here :))
